### PR TITLE
Chunked encoding for cluster state API

### PR DESCRIPTION
**Pull request by DaveCTurner**
Relates #89838

### CODE REVIEW

**Comment by DaveCTurner** [comment]
NB most of these changes are just to add support for XContent fragments throughout. Kind of noisy. I could make it a separate API instead of adding a boolean parameter, maybe that would be nicer.

**Comment by DaveCTurner** [comment]
https://github.com/elastic/elasticsearch/pull/92534 will fix the failure; @elasticmachine please run elasticsearch-ci/part-1

**Comment by DaveCTurner** [comment]
@original-brownbear I've merged this with `main`, adapting it to #92549, so this is worth another look

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -520,114 +519,135 @@ public String toString() {
...
+                    ChunkedToXContentHelper.endArray()
+                ),
+                (builder, params) -> builder.endObject().endObject()
+            ),
+
+            // customs
+            metrics.contains(Metric.CUSTOMS)
+                ? Iterators.flatMap(
+                    customs.entrySet().iterator(),
+                    cursor -> Iterators.concat(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java:643]: NIT: could use `org.elasticsearch.common.xcontent.ChunkedToXContentHelper#wrapWithObject` here to shorten this a little

File: server/src/main/java/org/elasticsearch/rest/action/DispatchingRestToXContentListener.java [hunk 2]
```diff
@@ -1,53 +0,0 @@
...
-import org.elasticsearch.xcontent.ToXContentObject;
-import org.elasticsearch.xcontent.XContentBuilder;
-
-import java.util.concurrent.ExecutorService;
-
-/**
- * Response listener for REST requests which dispatches the serialization of the response off of the thread on which the response was
- * received, since that thread is often a transport thread and XContent serialization might be expensive.
- */
-public class DispatchingRestToXContentListener<Response extends ToXContentObject> extends RestActionListener<Response> {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/rest/action/DispatchingRestToXContentListener.java:26]: :tada:

**Review by original-brownbear — APPROVED** [review]
Phew did my best to carefully read all the steps here, couldn't find any spot where the format would change or where we'd still serialize a problematically huge chunk => LGTM :)

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: ++ thanks, see c5bc46d577a

**Pull request merged**